### PR TITLE
drivers: tee: optee: Introduce SHM caching mechanism to the optee driver

### PR DESCRIPTION
**Pull request by oleksiimoisieiev**
This PR introduces the following features:
1) Enables shm_cache in OP-TEE to reduce SHM objects allocation. Decision whether to use shm_cache or not is done based on the capabilities exchange
2) SHM object cache which allow reusing SHM objects to store rpc args. Also concurrent calls can share SHM object memory by using different offsets. This allows to reduce alloc/free operation because for the most cases one SHM object is enogh.

### CODE REVIEW

**Review by firscity — COMMENTED** [review]
As I said, I have no objections from my side, but I would like to wait when @lorc take a look at this PR.

`Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>`

File: drivers/tee/optee/optee.c
Comment (xakep-amatop) [inline drivers/tee/optee/optee.c]:
```suggestion
	ROUND_UP(ROUND_UP(MAX_ARG_COUNT_PER_ENTRY, 8), sizeof(uint32_t))
```

File: drivers/tee/optee/optee.c
Comment (xakep-amatop) [inline drivers/tee/optee/optee.c]: violates misra rule 14.4

File: drivers/tee/optee/optee.c
Comment (xakep-amatop) [inline drivers/tee/optee/optee.c]: violates misra rule 14.4

File: drivers/tee/optee/optee.c
Comment (xakep-amatop) [inline drivers/tee/optee/optee.c]: violates misra rule 14.4
pls fix this one and all below violations

File: drivers/tee/optee/optee.c
Comment (xakep-amatop) [inline drivers/tee/optee/optee.c]: violates misra rule 10.1
pls, fix this and all below violations

File: drivers/tee/optee/optee.c
Comment (xakep-amatop) [inline drivers/tee/optee/optee.c]: violate misra rule 10.3

**Review by xakep-amatop — CHANGES_REQUESTED** [review]

File: drivers/tee/optee/Kconfig
Comment (xakep-amatop) [inline drivers/tee/optee/Kconfig]: by default all configs have disabled state
so, you can delete this line

File: drivers/tee/optee/optee.c
Comment (lorc) [inline drivers/tee/optee/optee.c]: Why you are enabling cache only if `rpc_param_count==0`? I see no such check in the linux kernel driver

File: drivers/tee/optee/optee.c
Comment (lorc) [inline drivers/tee/optee/optee.c:1411]: Could you please elaborate this? How it is possible that OP-TEE has cached shm buffers before driver is even being initialized?

File: drivers/tee/optee/optee.c
Comment (lorc) [inline drivers/tee/optee/optee.c]: No need to cast to `void*`

File: drivers/tee/optee/optee.c
Comment (lorc) [inline drivers/tee/optee/optee.c]: Why not `sys_bitarray_alloc(&entry->map, 1, &bit)` ?

File: drivers/tee/optee/optee.c
Comment (lorc) [inline drivers/tee/optee/optee.c]: It would be nice to add ASSERT just in case

File: drivers/tee/optee/optee.c
Comment (lorc) [inline drivers/tee/optee/optee.c]: It would be nice to add ASSERT just in case

**Comment by oleksiimoisieiev** [comment]
This MR is waiting until general OPTEE support will be merged to Zephyr master. https://github.com/zephyrproject-rtos/zephyr/pull/60680 
Right now TSC was passed okay. When the generic PR will be merged then this PR will be updated and ready to be merged

File: drivers/tee/optee/optee.c
Comment (Deedone) [reply drivers/tee/optee/optee.c]: Can you please elaborate on that?

File: drivers/tee/optee/optee.c
Comment (lorc) [reply drivers/tee/optee/optee.c:1411]: So, the question still stands. 

If this is a leftover from, say, U-Boot, then U-Boot should be fixed.

File: drivers/tee/optee/optee.c
Comment (lorc) [inline drivers/tee/optee/optee.c]: Outer parentheses are not necessary. This is not a macro

File: drivers/tee/optee/optee.c
Comment (lorc) [inline drivers/tee/optee/optee.c:748]: AFAIK, this is a relatively new feature. Not all OP-TEE versions support it (I am looking at you, Renesas). 
So you can't use it unconditionally. You need to check for `OPTEE_SMC_SEC_CAP_RPC_ARG` before using it.

File: drivers/tee/optee/optee.c
Comment (lorc) [inline drivers/tee/optee/optee.c]: function called `rm if empty`  but it calls `tee_rm_shm` unconditionally. I am not sure that I got the logic behind this

File: drivers/tee/optee/optee.c
Comment (lorc) [inline drivers/tee/optee/optee.c]: Well, `rm_shm_if_empty_unlocked` tries to remove `entry` unconditionally, if I got this right. What is the point of such "caching" then? Is it just to save RPC allocation trip? But RPC buffers are cached via different mechanism..

File: drivers/tee/optee/optee.c
Comment (oleksiimoisieiev) [reply drivers/tee/optee/optee.c:1411]: According to the comment from https://elixir.bootlin.com/linux/latest/source/drivers/tee/optee/smc_abi.c#L1749
There is a posibility that optee still has shm registered and no properly cleaned from the upper layer. I think this is not the case if zephyr is running as XEN domain, but I assume that there is a case when for example we start zephyr from u-boot after accessing optee from u-boot or other entity, such as older kernels, where resources are not properly cleaned.
I aware that it's unlikely to happen but this is some level of protection which is nice to have

Comment (lorc) [reply drivers/tee/optee/optee.c:1411]: Uhm, okay. I see. Anyways, it would be nice to have this explained in the comment.

**Comment by lorc** [comment]
Oh, and please fix "symultanious" spelling in the commit message for the last patch.

File: drivers/tee/optee/optee.c
Comment (Deedone) [reply drivers/tee/optee/optee.c:748]: It is checked indirectly through rpc_arg param, which depends on rpc_param_count, which depends on OPTEE_SMC_SEC_CAP_RPC_ARG. While it is a little convoluted I believe it's better to leave it like that to be more in tune with reference Linux implementation/

**Comment by Deedone** [comment]
Replaced the rm_if_empty approach with a hard limit of entries and updated the commit message of the last commit. Brought back the condition to enable optee_cache with better justification in the comments.

**Review by lorc — APPROVED** [review]
`Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

File: drivers/tee/optee/optee.c
Comment (oleksiimoisieiev) [inline drivers/tee/optee/optee.c]: hyprvisor -> hypervisor

**Pull request merged**
